src/resmom/start_exec.c

```diff
@@ -3566,7 +3566,7 @@ start_process(task *ptask, char **argv, char **envp, bool nodemux)
 	hook_output.progname = &progname;
 	CLEAR_HEAD(argv_list);
 	hook_output.argv = &argv_list;
-	hook_output.env = &env_str;
+	hook_output.env = NULL;
```

Comment:
The problem with the fix is that given an execjob_launch hook that specifies an environment variable change as in:

% cat launch.py

e=pbs.event()
e.env['HAPPY'] = "days"

# qmgr -c "c h launch event=execjob_launch"
# qmgr -c "i h launch application/x-python default launch.py"

Submitting an interactive job on a 2-node cluster:

% qsub -l select=2:ncpus=1 -I
pbsdsh -n 0 env
pbdsdsh -n 1 env

...
The output of 'pbsdsh' will show the HAPPY=days environment missing.

---
I'll attach the real fix to start_exec.c. There was a pull request that got merged awhile back dealing with this code, and it contained an incomplete fix. The things that went into finish_exec() did not make it to start_process(). Basically the fix diffs is as follows:

m> diff -c /tmp/start_exec.c.vaclav start_exec.c
*** /tmp/start_exec.c.vaclav    2016-11-17 12:14:09.855785938 -0500
--- start_exec.c        2016-11-17 15:58:36.136631439 -0500
***************
*** 98,106 ****

  #define EXTRA_ENV_PTRS               32

- /**
-  * @file
-  */
  /* Global Variables */

  extern        vnl_t           *vnlp;
--- 98,103 ----
***************
*** 1835,1841 ****
        int                     prolo_hooks = 0;/*# of runnable prologue hooks*/
        char                    *progname = NULL;
        pbs_list_head           argv_list;
-       char                    *env_str = NULL;
        char                    *the_progname;
        char                    **the_argv;
        char                    **the_env;
--- 1832,1837 ----
***************
*** 3280,3286 ****
        char                    hook_msg[HOOK_MSG_SIZE+1];
        char                    *progname = NULL;
        pbs_list_head           argv_list;
-       char                    *env_str = NULL;
        mom_hook_input_t        hook_input;
        mom_hook_output_t       hook_output;
        char                    *the_progname;
--- 3276,3281 ----
***************
*** 3566,3572 ****
        hook_output.progname = &progname;
        CLEAR_HEAD(argv_list);
        hook_output.argv = &argv_list;
-       hook_output.env = NULL;

        switch (mom_process_hooks(HOOK_EVENT_EXECJOB_LAUNCH,
                        PBS_MOM_SERVICE_NAME,
--- 3561,3566 ----
***************
*** 3577,3583 ****
                case 0: /* explicit reject */
                        free(progname);
                        free_attrlist(&argv_list);
!                       free(env_str);
                        starter_return(kid_write, kid_read,
                                        JOB_EXEC_FAILHOOK_DELETE, &sjr);
                case 1:   /* explicit accept */
--- 3571,3577 ----
                case 0: /* explicit reject */
                        free(progname);
                        free_attrlist(&argv_list);
!                       free_str_array(hook_output.env);
                        starter_return(kid_write, kid_read,
                                        JOB_EXEC_FAILHOOK_DELETE, &sjr);
                case 1:   /* explicit accept */
***************
*** 3591,3639 ****
                                        "execjob_launch hook returned NULL argv!");
                                free(progname);
                                free_attrlist(&argv_list);
!                               free(env_str);
                                starter_return(kid_write, kid_read,
                                        JOB_EXEC_FAILHOOK_DELETE, &sjr);
                        }
!                       if (env_str != NULL) {
!                               res_env = str_to_str_array(env_str, ",");
!                               if (res_env == NULL) {
!                                       log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_HOOK,
!                                               LOG_INFO, "",
!                                               "execjob_launch hook NULL env!");
!                                       free(progname);
!                                       free_attrlist(&argv_list);
!                                       free_str_array(the_argv);
!                                       free(env_str);
!                                       starter_return(kid_write, kid_read,
                                                JOB_EXEC_FAILHOOK_DELETE, &sjr);
!                               }

!                               /* clear the env array */
!                               vtable.v_used = 0;
!                               vtable.v_envp[0] = (char *)0;
!
!                               /* need to also set vtable as that would */
!                               /* get appended to later in the code */
!                               /* vtable holds the environmnent variables */
!                               /* and their values that are going to be */
!                               /* part of the job. */
!                               k = 0;
!                               while(res_env[k]) {
!                                       char    *n, *v, *p;
!                                       if ((p=strchr(res_env[k], '=')) !=
!                                                                       NULL) {
!                                               *p = '\0';
!                                               n = res_env[k];
!                                               v = p+1;
!                                               bld_env_variables(&vtable,
!                                                                       n, v);
!                                               *p = '=';
!                                       }
!                                       k++;
                                }
!                               the_env = vtable.v_envp;
                        }

                        break;
                case 2:   /* no hook script executed - go ahead and accept event */
--- 3585,3629 ----
                                        "execjob_launch hook returned NULL argv!");
                                free(progname);
                                free_attrlist(&argv_list);
!                               free_str_array(hook_output.env);
                                starter_return(kid_write, kid_read,
                                        JOB_EXEC_FAILHOOK_DELETE, &sjr);
                        }
!                       res_env = hook_output.env;
!                       if (res_env == NULL) {
!                               log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_HOOK,
!                                       LOG_INFO, "",
!                                       "execjob_launch hook NULL env!");
!                               free(progname);
!                               free_attrlist(&argv_list);
!                               free_str_array(the_argv);
!                               starter_return(kid_write, kid_read,
                                                JOB_EXEC_FAILHOOK_DELETE, &sjr);
!                       }

!                       /* clear the env array */
!                       vtable.v_used = 0;
!                       vtable.v_envp[0] = (char *)0;
!
!                       /* need to also set vtable as that would */
!                       /* get appended to later in the code */
!                       /* vtable holds the environmnent variables */
!                       /* and their values that are going to be */
!                       /* part of the job. */
!                       k = 0;
!                       while(res_env[k]) {
!                               char    *n, *v, *p;
!                               if ((p=strchr(res_env[k], '=')) != NULL) {
!                                       *p = '\0';
!                                       n = res_env[k];
!                                       v = p+1;
!                                       bld_env_variables(&vtable,
!                                                               n, v);
!                                       *p = '=';
                                }
!                               k++;
                        }
+                       the_env = vtable.v_envp;

                        break;
                case 2:   /* no hook script executed - go ahead and accept event */
***************
*** 3763,3769 ****
        free(progname);
        free_attrlist(&argv_list);
        free_str_array(the_argv);
!       free(env_str);
        free_str_array(the_env);
  #if 0
        /*
--- 3753,3759 ----
        free(progname);
        free_attrlist(&argv_list);
        free_str_array(the_argv);
!       free_str_array(hook_output.env);
        free_str_array(the_env);
  #if 0
        /*
***************
*** 5067,5073 ****
   *        create the file in its final destination directory and set the "keeping" flag so it will not be staged.
   *    5.  Else, the file path is created to put the file in PBS_HOME/spool
   #else
!  *    4.  Else, the file path is created to put the file in PBS_HOME/spool
   #endif localmod 118
   * @param[in]  pjob - pointer to job structure
   * @param[in]  which - identifies which file: StdOut, StdErr, or Chkpt.
--- 5057,5063 ----
   *        create the file in its final destination directory and set the "keeping" flag so it will not be staged.
   *    5.  Else, the file path is created to put the file in PBS_HOME/spool
   #else
!  *    6.  Else, the file path is created to put the file in PBS_HOME/spool
   #endif localmod 118
   * @param[in]  pjob - pointer to job structure
   * @param[in]  which - identifies which file: StdOut, StdErr, or Chkpt.
***************
*** 5774,5778 ****
        return(1);
  }
  #endif /* localmod 118 */
-
-
--- 5764,5766 ----

[start_exec.c.txt](https://github.com/PBSPro/pbspro/files/598483/start_exec.c.txt)